### PR TITLE
POLite software simulation

### PR DESCRIPTION
**Pull request by m8pple**
This is an attempt to pick just the software simulator out of my previous mess.

I've modified it a bit to try to make it as isolated as possible:

- I moved it into a subdir of `apps/POLite/util`, so that it is clear it is not an application.
- All the simulation logic is in a namespace `POLiteSWSim`, which resides in [`PGraph.h`](https://github.com/POETSII/tinsel/blob/dt10-polite-sw-sim/apps/POLite/util/POLiteSWSim/include/POLite/PGraph.h).
- The "fake" [`POLite.h`](https://github.com/POETSII/tinsel/blob/dt10-polite-sw-sim/apps/POLite/util/POLiteSWSim/include/POLite.h) just pulls a small number of types into the global namespace.
- `polite.mk` is modified to add `sim` and `sim-release` targets (I think I did slightly more than that, but the diff
   is not too bad I hope).

Apart from that, I also tried to make testing a bit clearer:

- New test script [`test_polite_sw_sim.sh`](https://github.com/POETSII/tinsel/blob/dt10-polite-sw-sim/apps/POLite/util/test_polite_sw_sim.sh) which tries to compile all the apps, and then run a sub-set
   which don't need input files.
- Updated the description of the readme.md to add a final `Simulation.` note to the [POLite section of
  the readme](https://github.com/POETSII/tinsel/blob/dt10-polite-sw-sim/README.md#9-polite-api).
- Added a [new appendix](https://github.com/POETSII/tinsel/blob/dt10-polite-sw-sim/README.md#j-testing) to the readme to talk about testing.

Unfortunately the actual simulation code in PGraph is still a giant blob, but I hope it is clear that it is additive
and shouldn't affect other parts of the repo, and at least there is testing there to show that it works.

I'd like to add more general testing in a separate pull, but kept it out of this one to minimise
the size of the patch.

### CODE REVIEW

**Comment by mn416** [comment]
Thanks @m8pple, I've made some tweaks to the docs.  I'm on holiday this week, so will aim to merge early next week after I've had chance to try it!

File: apps/POLite/util/polite.mk [hunk 1]
```diff
@@ -46,13 +52,39 @@ $(INC)/config.h: $(TINSEL_ROOT)/config.py
...
+	g++ -std=c++17 -g $(SIM_CPP_OBJ) -o $(BUILD)/sim -g -lmetis -ltbb -fopenmp
+
+$(BUILD)/%.sim-release.o : %.cpp
+	mkdir -p $(BUILD)
+	g++ -std=c++17 -g -c -O3 -DNDEBUG=1 -I $(POLITE_SW_SIM_DIR)/include -I $(INC) -I $(HL) -o $(BUILD)/$*.sim.o $*.cpp \
+		-fopenmp
 
-$(BUILD)/sim: $(RUN_CPP) $(RUN_H) $(HL)/sim/hostlink.a
-	g++ -O2 -I $(INC) -I $(HL) -o $(BUILD)/sim $(RUN_CPP) $(HL)/sim/hostlink.a \
-    -lmetis
```
Comment (mn416) [inline apps/POLite/util/polite.mk:55]: HI @m8pple, this target was/is useful for running POLite apps on top of the hardware (Bluesim) simulator.  Would be nice if this still works.  I don't mind this target is renamed (as it clashes with the new `sim` target for software simulation), but we probably don't want to remove it.

**Comment by mn416** [comment]
@m8pple Very nice, it  works for me, but please see the above comment about the hardware simulation target in the POLite Makefile before we can merge.

**Comment by mn416** [comment]
@m8pple, I've added back the POLite target for hardware simulation, renamed from `sim` to `hwsim` to avoid the clash.  Will merge soon, unless you have any final comments.

**Comment by m8pple** [comment]
Thanks.

I think I just deleted the sim target as a side-effect without thinking about it.

**Pull request merged**